### PR TITLE
[JENKINS-37241] Support for query parameters in autocomplete

### PR DESCRIPTION
**Pull request by timja**
See [JENKINS-37241](https://issues.jenkins.io/browse/JENKINS-37241).

### Testing done

Verified that existing usages in core are unaffected:

1. Creating a logger and having logger names suggested. This one is a good test as it doesn't have a descriptor behind it as well.
2. Copy from in a new job

Created a change in Azure VM agents https://github.com/jenkinsci/azure-vm-agents-plugin/pull/580 which uses the depends on functionality to access query parameters

### Proposed changelog entries

- Developer: Add support for `@QueryParameter` to the autocomplete component
- Change autocomplete component to use `POST` for sending requests

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/java/hudson/model/Descriptor.java [hunk 1]
```diff
@@ -474,6 +474,12 @@ public void calcAutoCompleteSettings(String field, Map<String, Object> attribute
         if (method == null)
             return;    // no auto-completion
 
+        // build query parameter line by figuring out what should be submitted
+        List<String> depends = buildFillDependencies(method, new ArrayList<>());
+        if (!depends.isEmpty()) {
+            attributes.put("fillDependsOn", String.join(" ", depends));
```
Comment (timja) [inline core/src/main/java/hudson/model/Descriptor.java:480]: I thought about whether this should be a different attribute name, e.g. `autocompleteFillDependsOn` but I couldn't see when they would ever clash so I just left it as the same name

File: src/main/js/components/dropdowns/autocomplete.js [hunk 2]
```diff
@@ -59,9 +59,30 @@ function init() {
...
+      depends.split(" ").forEach(
+        TryEach(function (n) {
+          q.nearBy(n);
+        }),
+      );
+    }
+
+    const queryString = q.toString();
+    const idx = queryString.indexOf("?");
+    const parameters = queryString.substring(idx + 1);
```
Comment (timja) [inline src/main/js/components/dropdowns/autocomplete.js:77]: copied from `hudson-behaviour.js` for the `doCheck` functionality. 

It uses query string by default and then if the method is post it converts to parameters, (which is the default)

I didn't see any reason for there to need to be an option for get, I'm not aware of it ever being used for the doChecks

**Comment by Kevin-CB** [comment]
Sorry for the delay! I’ll get started on the security review now.

**Comment by janfaracik** [comment]
Tested locally and LGTM. Seems there's quite a few competing components for this in Design Library, would be good for us to recommend one of them (and maybe dissuade usage of the others if they're not as functional).

**Review by janfaracik — APPROVED** [review]

**Review by janfaracik — APPROVED** [review]

**Review by krisstern — APPROVED** [review]
Makes sense

**Pull request merged**